### PR TITLE
Adding a new AdminRequest to query the health status of ambry server.

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Adding a new admin request for other party to query the status of ambry-server. 
Adding AdminResponseWithContent to carry richer response back to client.

### CODE REVIEW

**Review by SophieGuo410 — COMMENTED** [review]
Overall lgtm.

File: ambry-protocol/src/main/java/com/github/ambry/protocol/AdminResponseWithContent.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  /**
+   * Constructs an {@link AdminResponseWithContent}
+   * @param correlationId an ID to help match responses to requests.
+   * @param clientId the ID of the client.
+   * @param error the {@link ServerErrorCode} for the request.
+   * @param content the content to return back to client
+   */
+  public AdminResponseWithContent(int correlationId, String clientId, ServerErrorCode error, byte[] content) {
+    super(correlationId, clientId, ServerErrorCode.No_Error);
```
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-protocol/src/main/java/com/github/ambry/protocol/AdminResponseWithContent.java:48]: my mistake, will update

**Review by SophieGuo410 — APPROVED** [review]

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 2]
```diff
@@ -214,6 +214,15 @@ public void handleAdminRequest(NetworkRequest request) throws InterruptedExcepti
           requestTotalTimeHistogram = metrics.blobStoreControlRequestTotalTimeInMs;
           response = handleBlobStoreControlRequest(requestStream, adminRequest);
           break;
+        case HealthCheck:
+          metrics.blobStoreControlRequestQueueTimeInMs.update(requestQueueTime);
```
Comment (Arun-LinkedIn) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:218]: healthCheckRequestQueueTimeInMs, healthCheckRequestProcessingTimeInMs, etc

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 3]
```diff
@@ -399,6 +408,18 @@ private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStrea
...
+  /**
+   * Handles {@link com.github.ambry.protocol.AdminRequestOrResponseType#HealthCheck}.
+   * @param requestStream the serialized bytes of the request.
+   * @param adminRequest the {@link AdminRequest} received.
+   * @return the {@link AdminResponse} to the request.
+   * @throws IOException if there is any I/O error reading from the {@code requestStream}.
+   */
+  private AdminResponse handleHealthCheckRequest(DataInputStream requestStream, AdminRequest adminRequest)
+      throws StoreException, IOException {
+    return null;
```
Comment (Arun-LinkedIn) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:420]: QQ: Will this be implemented in next PR?

**Review by Arun-LinkedIn — APPROVED** [review]
Looks good to me too. Just one copy paste error needs correction.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:218]: updated.

**Pull request merged**